src/net/server/mod.rs

```diff
@@ -21,23 +27,71 @@ pub mod handlers;
 pub enum MpcServerError {
     #[error(transparent)]
     BadQueryString(#[from] QueryRejection),
+    #[error("header not found: {0}")]
+    MissingHeader(String),
+    #[error("invalid header: {0}")]
+    InvalidHeader(BoxError),
+    #[error(transparent)]
+    BadPathString(#[from] PathRejection),
+    #[error(transparent)]
+    HttpError(#[from] hyper::Error),
+    #[error("parse error: {0}")]
+    SerdeError(#[from] serde_json::Error),
+    #[error("could not forward messages: {0}")]
+    SendError(BoxError),
+}
+
+/// [`From`] implementation for [`MpcServerError::SendError`].
+/// first call `to_string` so as to drop `T` from the [`MpcServerError`]
+impl<T> From<mpsc::error::SendError<T>> for MpcServerError {
+    fn from(err: mpsc::error::SendError<T>) -> Self {
+        Self::SendError(err.to_string().into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::InvalidHeader`]
+impl From<std::num::ParseIntError> for MpcServerError {
+    fn from(err: std::num::ParseIntError) -> Self {
+        Self::InvalidHeader(err.into())
+    }
+}
+
+/// [`From`] implementation for [`MpcServerError::InvalidHeader`]
+impl From<axum::http::header::ToStrError> for MpcServerError {
+    fn from(err: axum::http::header::ToStrError) -> Self {
+        Self::InvalidHeader(err.into())
+    }
 }
 
 impl IntoResponse for MpcServerError {
     fn into_response(self) -> Response {
         let status_code = match &self {
-            MpcServerError::BadQueryString(_) => StatusCode::BAD_REQUEST,
+            Self::BadQueryString(_)
+            | Self::BadPathString(_)
+            | Self::SerdeError(_)
+            | Self::MissingHeader(_)
+            | Self::InvalidHeader(_) => StatusCode::BAD_REQUEST,
+            Self::HttpError(_) | Self::SendError(_) => StatusCode::INTERNAL_SERVER_ERROR,
         };
 
         (status_code, self.to_string()).into_response()
     }
 }
 
+/// TODO: how to remove <S, M> from router definition? It should not be there
```

Comment:
curious why do you need to get rid of `S`?

Reply:
ah, that was an old comment. let me remove the TODO